src/dataflow/src/render/context.rs

```diff
@@ -115,301 +99,369 @@ where
-                .collect::<Vec<_>>();
-            // Sort keys by length as a proxy for selectivity.
-            // TODO(#4580): improve this heuristic when possible.
-            constrained_keys.sort_by_key(|(key, _row)| key.len());
-            // If we found such a key, use a more advanced implementation.
-            if let Some((key, row)) = constrained_keys.pop() {
-                let (oks, errs) = &local[&key];
-                let result = Self::flat_map_ref_inner(oks, row, logic);
-                Some((result, errs.as_collection(|k, _v| k.clone())))
-            } else {
-                let (oks, errs) = local.values().next().expect("Empty arrangement");
-                Some((
-                    oks.flat_map_ref(move |_k, v| logic(RefOrMut::Ref(&v))),
-                    errs.as_collection(|k, _v| k.clone()),
-                ))
+        // If `key_val` is set, and we have the arrangement by that key, we should
+        // use that arrangement.
+        if let Some((key, val)) = key_val {
+            if let Some(flavor) = self.arrangement(&key) {
+                match flavor {
+                    ArrangementFlavor::Local(oks, errs) => {
+                        let oks = Self::flat_map_core(&oks, Some(val), logic);
+                        let errs = errs.as_collection(|k, _v| k.clone());
+                        return (oks, errs);
+                    }
+                    ArrangementFlavor::Trace(_, oks, errs) => {
+                        let oks = Self::flat_map_core(&oks, Some(val), logic);
+                        let errs = errs.as_collection(|k, _v| k.clone());
+                        return (oks, errs);
+                    }
+                }
             }
-        } else if let Some(trace) = self.trace.get(relation_expr) {
-            // Determine a key that is constrained to a literal value.
-            let mut constrained_keys = trace
-                .keys()
-                .filter_map(|key| key_selector(key).map(|row| (key.clone(), row)))
-                .collect::<Vec<_>>();
-            // Sort keys by length as a proxy for selectivity.
-            // TODO(#4580): improve this heuristic when possible.
-            constrained_keys.sort_by_key(|(key, _row)| key.len());
-            // If we found such a key, use a more advanced implementation.
-            if let Some((key, row)) = constrained_keys.pop() {
-                let (_id, oks, errs) = &trace[&key];
-                let result = Self::flat_map_ref_inner(oks, row, logic);
-                Some((result, errs.as_collection(|k, _v| k.clone())))
-            } else {
-                let (_id, oks, errs) = trace.values().next().expect("Empty arrangement");
-                Some((
-                    oks.flat_map_ref(move |_k, v| logic(RefOrMut::Ref(&v))),
-                    errs.as_collection(|k, _v| k.clone()),
-                ))
+        }
+
+        // No key, or no matching arrangement (that would be weird).
```

Comment:
Will update. It means we were asked to use a key expression and value, but there is no arrangement keyed by that expression. Probably worth logging an error up above as it would indicate a violated invariant.

Reply:
fixed